In [1]:
%matplotlib inline
import matplotlib.pyplot as plt
import nivapy3 as nivapy
import numpy as np
import pandas as pd
import seaborn as sn
import teotil2 as teo
import useful_rid_code as rid

sn.set_context("notebook")

# RID

## Estimating loads in unmonitored regions (parameterised)

This notebook is "parameterised" for use with Papermill. The cell below has the tag `parameters`, which means the entire notebook can be called from `01_recalculate_ospar_1990-2016_main.ipynb`.

The [TEOTIL2 model](https://nivanorge.github.io/teotil2/) is used to estimate loads in unmonitored areas. We know the regine ID for each of the 155 stations where water chemistry is measured, and we also know which OSPAR region each monitoring site drains to. We want to use observed data to estimate loads upstream of each monitoring point, and modelled data elsewhere.

In [2]:
# This cell is tagged 'parameters' for use with Papermill
# https://papermill.readthedocs.io/en/latest/index.html
year = 1990
user = ""
pw = ""

In [3]:
# Parameters
user = "jes"
pw = "BeakabusNov21.."
year = 2015


## 1. Generate model input file

In [4]:
# Parameters of interest
par_list = ["Tot-N", "Tot-P"]

# Path to TEOTIL2 "core" input data
teo_fold = r"../../../teotil2/data/core_input_data"

# Ouput path for model file
ann_input_csv = f"../../../teotil2/data/norway_annual_input_data/input_data_{year}.csv"

In [5]:
engine = rid.connect_to_nivabase(user=user, pw=pw)
df = teo.io.make_input_file(
    year, engine, teo_fold, ann_input_csv, mode="nutrients", par_list=par_list
)

Connection successful.


/opt/conda/lib/python3.9/site-packages/pandas/core/frame.py:5176: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(


## 2. Run model

In [6]:
%%time

# Run model
g = teo.model.run_model(ann_input_csv)

/opt/conda/lib/python3.9/site-packages/teotil2/model.py:67: FutureWarning: Boolean inputs to the `inclusive` argument are deprecated infavour of `both` or `neither`.
  df[f"trans_{par}"].between(0, 1, inclusive=True).all()


CPU times: user 6.69 s, sys: 125 ms, total: 6.81 s
Wall time: 6.87 s


## 3. Save results

In [7]:
# Save results as csv
res_csv = f"../../../teotil2/data/norway_annual_output_data/teotil2_results_{year}.csv"
df = teo.model.model_to_dataframe(g, out_path=res_csv)

df.head()

regine regine_ned  accum_agri_diff_tot-n_tonnes  \
0  001.1A2B   001.1A2A                      2.736090   
1  001.1A4D   001.1A4C                      0.000000   
2    001.1M     001.1L                      0.000000   
3  001.221Z   001.2210                      0.110388   
4  001.222Z   001.2220                      0.909652   

   accum_agri_diff_tot-p_tonnes  accum_agri_pt_tot-n_tonnes  \
0                      0.171273                    0.033277   
1                      0.000000                    0.000000   
2                      0.000000                    0.000000   
3                      0.006910                    0.001343   
4                      0.056942                    0.011063   

   accum_agri_pt_tot-p_tonnes  accum_all_point_tot-n_tonnes  \
0                    0.002690                      0.359071   
1                    0.000000                      0.000000   
2                    0.000000                      0.000000   
3                    0.000109                      0.014487   
4                    0.000894                      0.119378   

   accum_all_point_tot-p_tonnes  accum_all_sources_tot-n_tonnes  \
0                      0.027160                       15.501414   
1                      0.000000                        1.819635   
2                      0.000000                        4.196881   
3                      0.001096                        0.498729   
4                      0.009030                        2.465046   

   accum_all_sources_tot-p_tonnes  ...  local_ren_tot-n_tonnes  \
0                        0.542840  ...                     0.0   
1                        0.010962  ...                     0.0   
2                        0.036828  ...                     0.0   
3                        0.012064  ...                     0.0   
4                        0.082969  ...                     0.0   

   local_ren_tot-p_tonnes  local_runoff_mm/yr  local_spr_tot-n_tonnes  \
0                     0.0          430.723715                0.325794   
1                     0.0          502.511001                0.000000   
2                     0.0          502.511001                0.000000   
3                     0.0          430.723715                0.013144   
4                     0.0          430.723715                0.108315   

   local_spr_tot-p_tonnes  local_trans_tot-n  local_trans_tot-p  \
0                0.024470               1.00               1.00   
1                0.000000               0.81               0.26   
2                0.000000               0.86               0.31   
3                0.000987               1.00               1.00   
4                0.008135               1.00               1.00   

   local_urban_tot-n_tonnes  local_urban_tot-p_tonnes  local_vol_lake_m3  
0                   0.02205                   0.00315       1.788754e+08  
1                   0.00000                   0.00000       8.019497e+06  
2                   0.00000                   0.00000       3.866908e+07  
3                   0.00000                   0.00000       0.000000e+00  
4                   0.00000                   0.00000       0.000000e+00  

[5 rows x 54 columns]

In [8]:
# Save version with main catchments only
main_list = ["%03d." % i for i in range(1, 316)]
df2 = df.query("regine in @main_list")
df2.sort_values("regine", inplace=True)

# Save
main_csv = f"../../../Results/Unmon_loads/teotil2_results_{year}_main_catchs.csv"
df2.to_csv(main_csv, index=False, encoding="utf-8")

/opt/conda/lib/python3.9/site-packages/pandas/util/_decorators.py:311: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return func(*args, **kwargs)


## 4. Explore results

### 4.1. Total N and P

####  4.1.1. Identify areas with monitoring data

Where observations are available, we want to use them in preference to the model output. This means identifying all the catchments with observed data and substracting the model results for these locations. This is more complicated than it appears, because a small number of observed catchments are upstream of others, so subtracting all the loads for the 155 monitored catchments involves "double accounting", which we want to avoid. The first step is therefore to identify the downstream-most nodes for the monitored areas i.e. for the cases where one catchment is upstream of another, we just want the downstream node.

In [9]:
# Read station data
in_xlsx = r"../../../Data/RID_Sites_List_2017-2020.xlsx"
stn_df = pd.read_excel(in_xlsx, sheet_name="RID_All")
stn_df = stn_df.query("station_id != 38005")  # Ignore TROEMÅL2

# Get just cols of interest and drop duplicates
# (some sites are in the same regine)
stn_df = stn_df[["ospar_region", "nve_vassdrag_nr"]].drop_duplicates()

# Get catch IDs with obs data
obs_nds = set(stn_df["nve_vassdrag_nr"].values)

# Build network from input file
g, nd_list = teo.calib.build_calib_network(ann_input_csv, obs_nds)

# Get list of downstream nodes
ds_nds = []
for nd in g:
    # If no downstream nodes
    if g.out_degree(nd) == 0:
        # Node is of interest
        ds_nds.append(nd)

# Get just the downstream catchments
stn_df = stn_df[stn_df["nve_vassdrag_nr"].isin(ds_nds)]

#### 4.1.2. Sum model results for monitored locations

In [10]:
# Read model output
teo_df = pd.read_csv(res_csv)

# Join accumulated outputs to stns of interest
mon_df = pd.merge(
    stn_df, teo_df, how="left", left_on="nve_vassdrag_nr", right_on="regine"
)

# Groupby OSPAR region
mon_df = mon_df.groupby("ospar_region").sum()

# Get just accum cols
cols = [i for i in mon_df.columns if i.split("_")[0] == "accum"]
mon_df = mon_df[cols]

mon_df.head()

accum_agri_diff_tot-n_tonnes  \
ospar_region                                        
LOFOTEN-BARENTS SEA                    139.942219   
NORTH SEA                             2567.765031   
NORWEGIAN SEA2                        2672.211916   
SKAGERAK                             10868.081788   

                     accum_agri_diff_tot-p_tonnes  accum_agri_pt_tot-n_tonnes  \
ospar_region                                                                    
LOFOTEN-BARENTS SEA                      5.257379                    2.539554   
NORTH SEA                               67.413097                   34.862152   
NORWEGIAN SEA2                          71.211810                   40.763819   
SKAGERAK                               220.946016                  102.079627   

                     accum_agri_pt_tot-p_tonnes  accum_all_point_tot-n_tonnes  \
ospar_region                                                                    
LOFOTEN-BARENTS SEA                    0.218376                     86.653008   
NORTH SEA                              2.176470                    390.442614   
NORWEGIAN SEA2                         3.061545                    546.498244   
SKAGERAK                               5.546824                   3630.355202   

                     accum_all_point_tot-p_tonnes  \
ospar_region                                        
LOFOTEN-BARENTS SEA                      6.074953   
NORTH SEA                               36.296123   
NORWEGIAN SEA2                          41.075800   
SKAGERAK                                85.086804   

                     accum_all_sources_tot-n_tonnes  \
ospar_region                                          
LOFOTEN-BARENTS SEA                     3541.818152   
NORTH SEA                              12707.480632   
NORWEGIAN SEA2                         12267.687857   
SKAGERAK                               29800.268620   

                     accum_all_sources_tot-p_tonnes  \
ospar_region                                          
LOFOTEN-BARENTS SEA                       54.258047   
NORTH SEA                                184.308576   
NORWEGIAN SEA2                           237.916740   
SKAGERAK                                 456.438166   

                     accum_anth_diff_tot-n_tonnes  \
ospar_region                                        
LOFOTEN-BARENTS SEA                    139.942219   
NORTH SEA                             2573.737601   
NORWEGIAN SEA2                        2688.678894   
SKAGERAK                             11030.017027   

                     accum_anth_diff_tot-p_tonnes  ...  \
ospar_region                                       ...   
LOFOTEN-BARENTS SEA                      5.257379  ...   
NORTH SEA                               67.956573  ...   
NORWEGIAN SEA2                          73.340960  ...   
SKAGERAK                               238.423779  ...   

                     accum_nat_diff_tot-n_tonnes  accum_nat_diff_tot-p_tonnes  \
ospar_region                                                                    
LOFOTEN-BARENTS SEA                  3315.222924                    42.925715   
NORTH SEA                            9743.300417                    80.055881   
NORWEGIAN SEA2                       9032.510718                   123.499980   
SKAGERAK                            15139.896391                   132.927583   

                     accum_q_m3/s  accum_ren_tot-n_tonnes  \
ospar_region                                                
LOFOTEN-BARENTS SEA    742.520526               33.467862   
NORTH SEA             1450.241565              215.031310   
NORWEGIAN SEA2        1710.020502              307.243356   
SKAGERAK              1963.114038             2620.660200   

                     accum_ren_tot-p_tonnes  accum_spr_tot-n_tonnes  \
ospar_region                                                          
LOFOTEN-BARENTS SEA                1.197379               50.645592   
NORTH SEA   

This table gives the **modelled** inputs to each OSPAR region from catchments for which we have observed data. We want to subtract these values from the overall modelled inputs to each region and substitute the observed data instead.

The trickiest part of this is that the OSPAR regions in the TEOTIL catchment network (files named `regine_{year}.csv`) don't exactly match the relevant OSPAR definitions for this analysis. This is because the "OSPAR boundaries" in the model include catchments draining to Sweden (as part of TEOTIL2 Metals - see [here](https://nivanorge.github.io/teotil2/pages/07_1000_lakes.html)), so instead of using them directly we need to aggregate based on vassdragsnummers.

#### 4.1.3. Group model output according to "new" OSPAR regions

In [11]:
# Define "new" OSPAR regions (ranges are inclusive)
os_dict = {
    "SKAGERAK": (1, 23),
    "NORTH SEA": (24, 90),
    "NORWEGIAN SEA2": (91, 170),
    "LOFOTEN-BARENTS SEA": (171, 247),
}

# Container for results
df_list = []

# Loop over model output
for reg in os_dict.keys():
    min_id, max_id = os_dict[reg]

    regs = ["%03d." % i for i in range(min_id, max_id + 1)]

    # Get data for this region
    df2 = teo_df[teo_df["regine"].isin(regs)]

    # Get just accum cols
    cols = [i for i in df2.columns if i.split("_")[0] == "accum"]
    df2 = df2[cols]

    # Add region
    df2["ospar_region"] = reg

    # Add to output
    df_list.append(df2)

# Build df
os_df = pd.concat(df_list, axis=0)

# Aggregate
os_df = os_df.groupby("ospar_region").sum()

os_df.head()

accum_agri_diff_tot-n_tonnes  \
ospar_region                                        
LOFOTEN-BARENTS SEA                    585.201574   
NORTH SEA                             6182.495129   
NORWEGIAN SEA2                        7316.365359   
SKAGERAK                             12475.504474   

                     accum_agri_diff_tot-p_tonnes  accum_agri_pt_tot-n_tonnes  \
ospar_region                                                                    
LOFOTEN-BARENTS SEA                     27.562301                   10.565114   
NORTH SEA                              187.104456                   80.323926   
NORWEGIAN SEA2                         229.189818                  101.708615   
SKAGERAK                               280.250080                  112.985724   

                     accum_agri_pt_tot-p_tonnes  accum_all_point_tot-n_tonnes  \
ospar_region                                                                    
LOFOTEN-BARENTS SEA                    0.885819                  15439.673638   
NORTH SEA                              5.829535                  23988.602891   
NORWEGIAN SEA2                         8.082202                  28500.526686   
SKAGERAK                               6.352341                  10238.981092   

                     accum_all_point_tot-p_tonnes  \
ospar_region                                        
LOFOTEN-BARENTS SEA                   2606.916254   
NORTH SEA                             3793.059510   
NORWEGIAN SEA2                        4682.591366   
SKAGERAK                               260.460537   

                     accum_all_sources_tot-n_tonnes  \
ospar_region                                          
LOFOTEN-BARENTS SEA                    23968.262033   
NORTH SEA                              50964.914949   
NORWEGIAN SEA2                         52771.198715   
SKAGERAK                               39401.112857   

                     accum_all_sources_tot-p_tonnes  \
ospar_region                                          
LOFOTEN-BARENTS SEA                     2758.441928   
NORTH SEA                               4179.108941   
NORWEGIAN SEA2                          5161.971362   
SKAGERAK                                 719.677604   

                     accum_anth_diff_tot-n_tonnes  \
ospar_region                                        
LOFOTEN-BARENTS SEA                    595.173883   
NORTH SEA                             6222.312611   
NORWEGIAN SEA2                        7348.524149   
SKAGERAK                             12749.238943   

                     accum_anth_diff_tot-p_tonnes  ...  \
ospar_region                                       ...   
LOFOTEN-BARENTS SEA                     28.966142  ...   
NORTH SEA                              192.173126  ...   
NORWEGIAN SEA2                         233.430261  ...   
SKAGERAK                               312.982046  ...   

                     accum_nat_diff_tot-n_tonnes  accum_nat_diff_tot-p_tonnes  \
ospar_region                                                                    
LOFOTEN-BARENTS SEA                  7933.414512                   122.559532   
NORTH SEA                           20753.999447                   193.876305   
NORWEGIAN SEA2                      16922.147879                   245.949735   
SKAGERAK                            16412.892822                   146.235021   

                     accum_q_m3/s  accum_ren_tot-n_tonnes  \
ospar_region                                                
LOFOTEN-BARENTS SEA   1991.626067              982.028835   
NORTH SEA             3201.924379             3560.554520   
NORWEGIAN SEA2        3422.183045             2560.436296   
SKAGERAK              2092.583595             8054.283449   

                     accum_ren_tot-p_tonnes  accum_spr_tot-n_tonnes  \
ospar_region                                                          
LOFOTEN-BARENTS SEA              131.575975              305.310495   
NORTH SEA   

We can now calculate the unmonitored component by simply subtracting the values modelled upstream of monitoring stations from the overall modelled inputs to each OSPAR region.

#### 4.1.4. Estimate loads in unmonitored areas

In [12]:
# Calc unmonitored loads
unmon_df = os_df - mon_df

# Write output
out_csv = f"../../../Results/Unmon_loads/teotil2_raw_unmonitored_loads_{year}.csv"
unmon_df.to_csv(out_csv, encoding="utf-8", index_label="ospar_region")

unmon_df.round(0).astype(int).T

ospar_region                    LOFOTEN-BARENTS SEA  NORTH SEA  \
accum_agri_diff_tot-n_tonnes                    445       3615   
accum_agri_diff_tot-p_tonnes                     22        120   
accum_agri_pt_tot-n_tonnes                        8         45   
accum_agri_pt_tot-p_tonnes                        1          4   
accum_all_point_tot-n_tonnes                  15353      23598   
accum_all_point_tot-p_tonnes                   2601       3757   
accum_all_sources_tot-n_tonnes                20426      38257   
accum_all_sources_tot-p_tonnes                 2704       3995   
accum_anth_diff_tot-n_tonnes                    455       3649   
accum_anth_diff_tot-p_tonnes                     24        124   
accum_aqu_tot-n_tonnes                        14083      19268   
accum_aqu_tot-p_tonnes                         2433       3324   
accum_ind_tot-n_tonnes                           58        400   
accum_ind_tot-p_tonnes                            4         83   
accum_nat_diff_tot-n_tonnes                    4618      11011   
accum_nat_diff_tot-p_tonnes                      80        114   
accum_q_m3/s                                   1249       1752   
accum_ren_tot-n_tonnes                          949       3346   
accum_ren_tot-p_tonnes                          130        279   
accum_spr_tot-n_tonnes                          255        538   
accum_spr_tot-p_tonnes                           33         67   
accum_upstr_area_km2                          74535      35961   
accum_urban_tot-n_tonnes                         10         34   
accum_urban_tot-p_tonnes                          1          5   

ospar_region                    NORWEGIAN SEA2  SKAGERAK  
accum_agri_diff_tot-n_tonnes              4644      1607  
accum_agri_diff_tot-p_tonnes               158        59  
accum_agri_pt_tot-n_tonnes                  61        11  
accum_agri_pt_tot-p_tonnes                   5         1  
accum_all_point_tot-n_tonnes             27954      6609  
accum_all_point_tot-p_tonnes              4642       175  
accum_all_sources_tot-n_tonnes           40504      9601  
accum_all_sources_tot-p_tonnes            4924       263  
accum_anth_diff_tot-n_tonnes              4660      1719  
accum_anth_diff_tot-p_tonnes               160        75  
accum_aqu_tot-n_tonnes                   24174        43  
accum_aqu_tot-p_tonnes                    4184         8  
accum_ind_tot-n_tonnes                     930       954  
accum_ind_tot-p_tonnes                      87        34  
accum_nat_diff_tot-n_tonnes               7890      1273  
accum_nat_diff_tot-p_tonnes                122        13  
accum_q_m3/s                              1712       129  
accum_ren_tot-n_tonnes                    2253      5434  
accum_ren_tot-p_tonnes                     298       116  
accum_spr_tot-n_tonnes                     536       167  
accum_spr_tot-p_tonnes                      68        18  
accum_upstr_area_km2                     68037      8629  
accum_urban_tot-n_tonnes                    16       112  
accum_urban_tot-p_tonnes                     2        15

#### 4.1.5. Aggregate values to required quantities

In [13]:
# Aggregate to match report
unmon_df["flow"] = unmon_df["accum_q_m3/s"] * 60 * 60 * 24 / 1000.0  # 1000s m3/day

unmon_df["sew_n"] = (
    unmon_df["accum_ren_tot-n_tonnes"] + unmon_df["accum_spr_tot-n_tonnes"]
)
unmon_df["sew_p"] = (
    unmon_df["accum_ren_tot-p_tonnes"] + unmon_df["accum_spr_tot-p_tonnes"]
)

unmon_df["ind_n"] = unmon_df["accum_ind_tot-n_tonnes"]
unmon_df["ind_p"] = unmon_df["accum_ind_tot-p_tonnes"]

unmon_df["fish_n"] = unmon_df["accum_aqu_tot-n_tonnes"]
unmon_df["fish_p"] = unmon_df["accum_aqu_tot-p_tonnes"]

unmon_df["diff_n"] = (
    unmon_df["accum_anth_diff_tot-n_tonnes"] + unmon_df["accum_nat_diff_tot-n_tonnes"]
)
unmon_df["diff_p"] = (
    unmon_df["accum_anth_diff_tot-p_tonnes"] + unmon_df["accum_nat_diff_tot-p_tonnes"]
)

new_df = unmon_df[
    ["flow", "sew_n", "sew_p", "ind_n", "ind_p", "fish_n", "fish_p", "diff_n", "diff_p"]
]

# Total for Norway
new_df.loc["NORWAY"] = new_df.sum(axis=0)

# Reorder rows
new_df = new_df.reindex(
    ["NORWAY", "LOFOTEN-BARENTS SEA", "NORTH SEA", "NORWEGIAN SEA2", "SKAGERAK"]
)

new_df.round().astype(int)

/opt/conda/lib/python3.9/site-packages/pandas/core/indexing.py:723: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value, self.name)


flow  sew_n  sew_p  ind_n  ind_p  fish_n  fish_p  \
ospar_region                                                              
NORWAY               418385  13478   1008   2342    208   57569    9948   
LOFOTEN-BARENTS SEA  107923   1203    163     58      4   14083    2433   
NORTH SEA            151345   3884    346    400     83   19268    3324   
NORWEGIAN SEA2       147931   2790    366    930     87   24174    4184   
SKAGERAK              11186   5601    133    954     34      43       8   

                     diff_n  diff_p  
ospar_region                         
NORWAY                35274     712  
LOFOTEN-BARENTS SEA    5073     103  
NORTH SEA             14659     238  
NORWEGIAN SEA2        12549     283  
SKAGERAK               2992      88

## 5. Other N and P species

Tore's procedure `RESA2.FIXTEOTILPN` defines simple correction factors for estimating PO4, NO3 and NH4 from total P and N. The table below lists the factors used.

|   Source    | Phosphate | Nitrate | Ammonium |
|:-----------:|:---------:|:-------:|:--------:|
|    Sewage   |     0.600 |   0.050 |    0.750 |
|   Industry  |     0.600 |   0.050 |    0.750 |
| Aquaculture |     0.690 |   0.110 |    0.800 |
|   Diffuse   |     0.246 |   0.625 |    0.055 |


In [14]:
# Dict of conversion factors
con_dict = {
    ("sew", "po4"): ("p", 0.6),
    ("ind", "po4"): ("p", 0.6),
    ("fish", "po4"): ("p", 0.69),
    ("diff", "po4"): ("p", 0.246),
    ("sew", "no3"): ("n", 0.05),
    ("ind", "no3"): ("n", 0.05),
    ("fish", "no3"): ("n", 0.11),
    ("diff", "no3"): ("n", 0.625),
    ("sew", "nh4"): ("n", 0.75),
    ("ind", "nh4"): ("n", 0.75),
    ("fish", "nh4"): ("n", 0.8),
    ("diff", "nh4"): ("n", 0.055),
}

# Apply factors
for src in ["sew", "ind", "fish", "diff"]:
    for spc in ["po4", "no3", "nh4"]:
        el, fac = con_dict[(src, spc)]
        new_df[src + "_" + spc] = fac * new_df[src + "_" + el]

new_df.round().astype(int).T

ospar_region  NORWAY  LOFOTEN-BARENTS SEA  NORTH SEA  NORWEGIAN SEA2  SKAGERAK
flow          418385               107923     151345          147931     11186
sew_n          13478                 1203       3884            2790      5601
sew_p           1008                  163        346             366       133
ind_n           2342                   58        400             930       954
ind_p            208                    4         83              87        34
fish_n         57569                14083      19268           24174        43
fish_p          9948                 2433       3324            4184         8
diff_n         35274                 5073      14659           12549      2992
diff_p           712                  103        238             283        88
sew_po4          605                   98        208             219        80
sew_no3          674                   60        194             139       280
sew_nh4        10108                  902       2913            2092      4201
ind_po4          125                    3         50              52        20
ind_no3          117                    3         20              46        48
ind_nh4         1757                   44        300             697       715
fish_po4        6864                 1679       2293            2887         5
fish_no3        6333                 1549       2120            2659         5
fish_nh4       46055                11267      15415           19339        35
diff_po4         175                   25         59              70        22
diff_no3       22046                 3171       9162            7843      1870
diff_nh4        1940                  279        806             690       165

## 6. Other quantities

The model currently only considers N and P, but the project focuses on a wider range of parameters. For now, we simply assume that all measured inputs (`renseanlegg`, `industri` and `akvakultur`) for regines outside of catchments with measured data make it to the sea.

We only want data for catchments that are not monitored i.e. for regine IDs **not** in the graph created above.

In [15]:
engine = rid.connect_to_nivabase(user=user, pw=pw)

# The sql below uses a horrible (and slow!) hack to get around Oracle's
# 1000 item limit on IN clauses. See here for details:
# https://stackoverflow.com/a/9084247/505698
nd_list_hack = [(1, i) for i in nd_list]

sql = (
    "SELECT SUBSTR(a.regine, 1, 3) AS vassdrag, "
    "  a.type, "
    "  b.name, "
    "  b.unit, "
    "  SUM(c.value * d.factor) as value "
    "FROM RESA2.RID_PUNKTKILDER a, "
    "RESA2.RID_PUNKTKILDER_OUTPAR_DEF b, "
    "RESA2.RID_PUNKTKILDER_INPAR_VALUES c, "
    "RESA2.RID_PUNKTKILDER_INP_OUTP d "
    "WHERE a.anlegg_nr = c.anlegg_nr "
    "AND (1, a.regine) NOT IN %s "
    "AND d.in_pid = c.inp_par_id "
    "AND d.out_pid = b.out_pid "
    "AND c.year = %s "
    "GROUP BY SUBSTR(a.regine, 1, 3), a.type, b.name, b.unit "
    "ORDER BY SUBSTR(a.regine, 1, 3), a.type" % (tuple(nd_list_hack), year)
)

df = pd.read_sql(sql, engine)

# Tidy
df["par"] = df["type"] + "_" + df["name"] + "_" + df["unit"]
del df["name"], df["unit"], df["type"]

# Pivot
df = df.pivot(index="vassdrag", columns="par", values="value")
df.reset_index(inplace=True)

Connection successful.


In [16]:
def f(x):
    try:
        a = int(x)
        return a
    except:
        return -999


# Convert vassdrag to numbers
df["vass"] = df["vassdrag"].apply(f)

# Get just the main catchments
df = df.query("vass != -999")

df.head()

par vassdrag  INDUSTRI_As_tonn  INDUSTRI_Cd_tonn  INDUSTRI_Cr_tonn  \
0        001            0.0597            0.0052            0.0105   
1        002            0.0133            0.0055            0.6729   
2        003               NaN               NaN               NaN   
3        004               NaN               NaN               NaN   
4        005               NaN               NaN               NaN   

par  INDUSTRI_Cu_tonn  INDUSTRI_Hg_tonn  INDUSTRI_NH3_tonn  \
0              0.0470               NaN                NaN   
1              3.3083             0.012                NaN   
2              0.0004               NaN                NaN   
3                 NaN               NaN                NaN   
4                 NaN               NaN                NaN   

par  INDUSTRI_NH4-N_tonn  INDUSTRI_Ni_tonn  INDUSTRI_PAH_tonn  ...  \
0                    NaN            0.0494                NaN  ...   
1                    NaN            0.4234             0.0002  ...   
2                    NaN            0.0002                NaN  ...   
3                    NaN               NaN                NaN  ...   
4                    NaN               NaN                NaN  ...   

par  RENSEANLEGG_Hg_tonn  RENSEANLEGG_Ni_tonn  RENSEANLEGG_PAH_tonn  \
0               0.000048              0.02807              0.000600   
1               0.000147              0.01815              0.000085   
2                    NaN                  NaN                   NaN   
3                    NaN                  NaN                   NaN   
4                    NaN                  NaN                   NaN   

par  RENSEANLEGG_PCB_tonn  RENSEANLEGG_Pb_tonn  RENSEANLEGG_S.P.M._tonn  \
0                0.000000              0.00494                      NaN   
1                0.000051              0.00448                    0.025   
2                     NaN                  NaN                      NaN   
3                     NaN                  NaN                      NaN   
4                     NaN                  NaN                      NaN   

par  RENSEANLEGG_Tot-N_tonn  RENSEANLEGG_Tot-P_tonn  RENSEANLEGG_Zn_tonn  vass  
0                 110.98735                 5.13800              0.14505     1  
1                 687.16123                17.36042              0.18853     2  
2                 282.00000                 2.81000                  NaN     3  
3                 142.99000                 1.69000                  NaN     4  
4                  32.72333                 1.34420                  NaN     5  

[5 rows x 31 columns]

In [17]:
def f2(x):
    if x in range(1, 24):
        return "SKAGERAK"
    elif x in range(24, 91):
        return "NORTH SEA"
    elif x in range(91, 171):
        return "NORWEGIAN SEA2"
    elif x in range(171, 248):
        return "LOFOTEN-BARENTS SEA"
    else:
        return np.nan


# Assign main catchments to OSPAR regions
df["osp_reg"] = df["vass"].apply(f2)

df.head()

par vassdrag  INDUSTRI_As_tonn  INDUSTRI_Cd_tonn  INDUSTRI_Cr_tonn  \
0        001            0.0597            0.0052            0.0105   
1        002            0.0133            0.0055            0.6729   
2        003               NaN               NaN               NaN   
3        004               NaN               NaN               NaN   
4        005               NaN               NaN               NaN   

par  INDUSTRI_Cu_tonn  INDUSTRI_Hg_tonn  INDUSTRI_NH3_tonn  \
0              0.0470               NaN                NaN   
1              3.3083             0.012                NaN   
2              0.0004               NaN                NaN   
3                 NaN               NaN                NaN   
4                 NaN               NaN                NaN   

par  INDUSTRI_NH4-N_tonn  INDUSTRI_Ni_tonn  INDUSTRI_PAH_tonn  ...  \
0                    NaN            0.0494                NaN  ...   
1                    NaN            0.4234             0.0002  ...   
2                    NaN            0.0002                NaN  ...   
3                    NaN               NaN                NaN  ...   
4                    NaN               NaN                NaN  ...   

par  RENSEANLEGG_Ni_tonn  RENSEANLEGG_PAH_tonn  RENSEANLEGG_PCB_tonn  \
0                0.02807              0.000600              0.000000   
1                0.01815              0.000085              0.000051   
2                    NaN                   NaN                   NaN   
3                    NaN                   NaN                   NaN   
4                    NaN                   NaN                   NaN   

par  RENSEANLEGG_Pb_tonn  RENSEANLEGG_S.P.M._tonn  RENSEANLEGG_Tot-N_tonn  \
0                0.00494                      NaN               110.98735   
1                0.00448                    0.025               687.16123   
2                    NaN                      NaN               282.00000   
3                    NaN                      NaN               142.99000   
4                    NaN                      NaN                32.72333   

par  RENSEANLEGG_Tot-P_tonn  RENSEANLEGG_Zn_tonn  vass   osp_reg  
0                   5.13800              0.14505     1  SKAGERAK  
1                  17.36042              0.18853     2  SKAGERAK  
2                   2.81000                  NaN     3  SKAGERAK  
3                   1.69000                  NaN     4  SKAGERAK  
4                   1.34420                  NaN     5  SKAGERAK  

[5 rows x 32 columns]

In [18]:
# Group by OSPAR region
df.fillna(0, inplace=True)
df = df.groupby("osp_reg").sum()
if 0 in df.index:
    df.drop(0, inplace=True)

# Total for Norway
df.loc["NORWAY"] = df.sum(axis=0)

# Join to model results
df = new_df.join(df)

# Get cols of interest
umod_cols = ["S.P.M.", "TOC", "As", "Pb", "Cd", "Cu", "Zn", "Ni", "Cr", "Hg"]
umod_cols = [
    "%s_%s_tonn" % (i, j) for i in ["INDUSTRI", "RENSEANLEGG"] for j in umod_cols
]
cols = list(new_df.columns) + umod_cols
cols.remove("RENSEANLEGG_TOC_tonn")
cols = [i for i in cols if i in df.columns]
df = df[cols]

df.round(0).astype(int).T

ospar_region             NORWAY  LOFOTEN-BARENTS SEA  NORTH SEA  \
flow                     418385               107923     151345   
sew_n                     13478                 1203       3884   
sew_p                      1008                  163        346   
ind_n                      2342                   58        400   
ind_p                       208                    4         83   
fish_n                    57569                14083      19268   
fish_p                     9948                 2433       3324   
diff_n                    35274                 5073      14659   
diff_p                      712                  103        238   
sew_po4                     605                   98        208   
sew_no3                     674                   60        194   
sew_nh4                   10108                  902       2913   
ind_po4                     125                    3         50   
ind_no3                     117                    3         20   
ind_nh4                    1757                   44        300   
fish_po4                   6864                 1679       2293   
fish_no3                   6333                 1549       2120   
fish_nh4                  46055                11267      15415   
diff_po4                    175                   25         59   
diff_no3                  22046                 3171       9162   
diff_nh4                   1940                  279        806   
INDUSTRI_S.P.M._tonn     672649               228798     130162   
INDUSTRI_TOC_tonn           749                  138        386   
INDUSTRI_As_tonn              5                    0          5   
INDUSTRI_Pb_tonn              1                    0          1   
INDUSTRI_Cd_tonn              0                    0          0   
INDUSTRI_Cu_tonn              5                    0          1   
INDUSTRI_Zn_tonn             17                    0          8   
INDUSTRI_Ni_tonn              6                    0          4   
INDUSTRI_Cr_tonn              1                    0          0   
INDUSTRI_Hg_tonn              0                    0          0   
RENSEANLEGG_S.P.M._tonn    2010                  262        877   
RENSEANLEGG_As_tonn           0                    0          0   
RENSEANLEGG_Pb_tonn           0                    0          0   
RENSEANLEGG_Cd_tonn           0                    0          0   
RENSEANLEGG_Cu_tonn           4                    0          1   
RENSEANLEGG_Zn_tonn          12                    0          2   
RENSEANLEGG_Ni_tonn           2                    0          0   
RENSEANLEGG_Cr_tonn           1                    0          0   
RENSEANLEGG_Hg_tonn           0                    0          0   

ospar_region             NORWEGIAN SEA2  SKAGERAK  
flow                             147931     11186  
sew_n                              2790      5601  
sew_p                               366       133  
ind_n                               930       954  
ind_p                                87        34  
fish_n                            24174        43  
fish_p                             4184         8  
diff_n                            12549      2992  
diff_p                              283        88  
sew_po4                             219        80  
sew_no3                             139       280  
sew_nh4                            2092      4201  
ind_po4                              52        20  
ind_no3                              46        48  
ind_nh4                             697       715  
fish_po4                           2887         5  
fish_no3                           2659         5  
fish_nh4                          19339        35  
diff_po4                             70        22  
diff_no3                           7843      1870  
diff_nh4                            690       165  
INDUSTRI_S.P.M._tonn             312038      1651  
INDUSTRI_TOC_tonn                   148        78  
INDU

## 7. Fish farm copper

Finally, we need to add in the Cu totals from fish farms. The method is similar to that used above, but simpler because we're only interested in one parameter.

In [19]:
engine = rid.connect_to_nivabase(user=user, pw=pw)

# The sql below uses a horrible (and slow!) hack to get around Oracle's
# 1000 item limit on IN clauses. See here for details:
# https://stackoverflow.com/a/9084247/505698
nd_list_hack = [(1, i) for i in nd_list]

sql = (
    "SELECT SUBSTR(regine, 1, 3) AS vassdrag, "
    "  SUM(value) AS value FROM ( "
    "    SELECT b.regine, "
    "           c.name, "
    "           (a.value*d.factor) AS value "
    "    FROM resa2.rid_kilder_aqkult_values a, "
    "    resa2.rid_kilder_aquakultur b, "
    "    resa2.rid_punktkilder_outpar_def c, "
    "    resa2.rid_punktkilder_inp_outp d "
    "    WHERE a.anlegg_nr = b.nr "
    "    AND (1, b.regine) NOT IN %s "
    "    AND a.inp_par_id = d.in_pid "
    "    AND c.out_pid = d.out_pid "
    "    AND name = 'Cu' "
    "    AND ar = %s) "
    "GROUP BY SUBSTR(regine, 1, 3)" % (tuple(nd_list_hack), year)
)

aq_df = pd.read_sql(sql, engine)

if len(aq_df) > 0:
    # Get vassdrag
    aq_df["vass"] = aq_df["vassdrag"].apply(f)
    aq_df = aq_df.query("vass != -999")

    # Calc OSPAR region and group
    aq_df["osp_reg"] = aq_df["vass"].apply(f2)
    aq_df.fillna(0, inplace=True)
    aq_df = aq_df.groupby("osp_reg").sum()
    del aq_df["vass"]

    # Total for Norway
    aq_df.loc["NORWAY"] = aq_df.sum(axis=0)

    # Rename
    aq_df.columns = [
        "AQUAKULTUR_Cu_tonn",
    ]

    # Join model results
    df = df.join(aq_df)

    df.round(0).astype(int).T

Connection successful.


In [20]:
# Write output
out_csv = f"../../../Results/Unmon_loads/teotil2_ospar_unmonitored_loads_{year}.csv"
df.to_csv(out_csv)

This data can then be used to create Table 3 in the report - see [this notebook](https://nbviewer.jupyter.org/github/JamesSample/rid/blob/master/notebooks/summary_table_2017.ipynb) for details.